### ISSUE TITLE
Design Proposal: Processes as Entities

### ISSUE THREAD

**Issue by TimJentzsch** [edouardpoitras/bevy_local_commands#1]
I was planning to develop a similar crate as this one and therefore thought a lot about potential API designs.

One design that I think is worth exploring is to treat processes as entities in Bevy.
So to spawn a process, you would spawn an entity with a particular component (e.g. `Process` or `Command` component) and to kill a process you would despawn this entity.
The plugin's systems would take care of creating the actual processes based on the `Created` events and kill the process (or if the process ends it would despawn the entity).

I think this design provides a couple of advantages:

- It uses Bevy-native concepts like spawning and despawning entities.
- If you have many different processes, you can categorize and filter them with marker components, etc. You can use the query mechanisms to organize your processes.
- You can attach additional data to the processes simply by adding additional components to the entities.
- Looking up specific processes becomes faster, as lookup by entity ID is efficient in Bevy. The PID can still be provided in the form of a component attached to the entity.

Of course there may also be some disadvantages that I haven't thought of yet. For example, despawning the entity once the process has finished might be harder than I think.
Still, I think it might be worth it to explore this design.

**Comment by edouardpoitras** [comment]
I think that's a solid idea. I also wanted to explore processes as entities in the ECS - the current implementation is just a dumb approach to get something up-and-running for my other project.

Some things to consider:
- Receiving/parsing process output
    - Do we attach the output to the entity?
    - Buffer size and limit? (this is currently done by sending events of buffered output every second)
- Killing processes
    - Doesn't work very well because the process/task blocks when waiting on command output
    - This means the actual duct kill command won't execute until output is received
    - Tested task.cancel() and task.abort() with no luck
    - Can consider killing from outside the task?
- Chaining executions
    - A common use-case seems to be chaining these commands one at a time until one fails or they all complete successfully
    - Would be nice to have this built-into the library

I'm down to give it a shot. Feel free to start up a branch/PR. Otherwise I may get around to it at a later date.

Edit: Let's try to keep a simple event-based API (maybe it just uses the ECS under the hood), for one-off commands that you don't care about input/output. I like events :)

**Comment by edouardpoitras** [comment]
I'm also not convinced that Arc<Mutex<T>> is the way to go here. It's kind of ugly. Channels may be a better approach for communicating between the threads.
